### PR TITLE
Fix property access on an object literal

### LINKED ISSUE #10129: Object literal cast compiles to produce JS with syntax error
**TypeScript Version:**  2.0.0

**Code**

``` ts
class A { }

window.onload = () => {
    (<A>{}).toString();
};
```

**Expected behavior:** The toString line should compile to: `({}).toString()` or `({}.toString())` and hence there should not be a runtime syntax error

**Actual behavior:** The toString line compiles to `{}.toString()`, which is incorrect syntax and hence a syntax error is thrown at runtime.

### PR DESCRIPTION
**Pull request by henrymercer**
Fixes #10129.  

Currently a property access on an object literal e.g. `{ a : 1 }` can compile to `{ a : 1 }.x`, which is incorrect syntax.  This fixes object literal property accesses to compile to `({ a : 1 }.x)`, and fixes existing tests that have been using this incorrect syntax.

### CODE REVIEW

**Comment by SaschaNaz** [comment]
Why only empty object literal? `(<any>{a:3}).toString()` also causes the issue.

File: src/compiler/factory.ts [hunk 1]
```diff
@@ -3811,14 +3811,18 @@ namespace ts {
...
         //       new C.x        -> not the same as (new C).x
         //
+        //    ObjectLiteral:
+        //       {a:1}.toString()  -> is incorrect syntax, should be ({a:1}).toString()
+        //
         const emittedExpression = skipPartiallyEmittedExpressions(expression);
         if (isLeftHandSideExpression(emittedExpression)
-            && (emittedExpression.kind !== SyntaxKind.NewExpression || (<NewExpression>emittedExpression).arguments)) {
+            && (!isNewExpression(emittedExpression) || (<NewExpression>emittedExpression).arguments)
+            && !isObjectLiteralExpression(emittedExpression)) {
```
Comment (mhegazy) [inline src/compiler/factory.ts]: this is only true if this is an expression statement. i.e.

```ts
var x = {a: 1}.a;   // OK

{a: 1}.a;   // Error
```

File: tests/baselines/reference/assignmentTypeNarrowing.js [hunk 2]
```diff
@@ -37,13 +37,13 @@ x = [true][0];
 x; // boolean
 _a = [1][0], x = _a === void 0 ? "" : _a;
 x; // string | number
-(x = { x: true }.x);
+(x = ({ x: true }).x);
```
Comment (rbuckton) [inline tests/baselines/reference/assignmentTypeNarrowing.js]: The parenthesis aren't needed here.

**Comment by henrymercer** [comment]
Thanks @rbuckton, adding the parentheses around the whole property access expression in `parenthesizeExpressionForExpressionStatement` also fixes the problem and is much less aggressive in doing so.  I have implemented the fix as you suggested.

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**